utils/symbol.c

```diff
@@ -1430,3 +1423,61 @@ void print_symtabs(struct symtabs *symtabs)
 		symbol_putname(sym, name);
 	}
 }
+
+static unsigned long kernel_base_addr;
+
+static unsigned long get_kernel_base(char *str)
+{
+	unsigned long addr = strtoul(str, NULL, 16);
+
+	if (addr < 0x40000000UL) {
+		return 0x40000000UL;
+	} else if (addr < 0x80000000UL) {
+		return 0x80000000UL;
+	} else if (addr < 0xB0000000UL) {
+		return 0xB0000000UL;
+	} else if (addr < 0xC0000000UL) {
+		return 0xC0000000UL;
+	} else {
+		return 0x800000000000ULL;
+	}
+}
+
+void set_kernel_base(char *dirname, const char *session_id)
+{
+	FILE *fp;
+	char buf[4096];
+	char line[200];
+
+	snprintf(buf, sizeof(buf), "%s/sid-%s.map", dirname, session_id);
+
+	fp = fopen(buf, "r");
+	if (fp == NULL) {
+		if (errno == EEXIST)
+			return;
+		pr_err("cannot open %s file", buf);
+	}
+
+	while (fgets(line, sizeof(line), fp) != NULL) {
+		if (strstr(line, "[stack") != NULL) {
+			kernel_base_addr = get_kernel_base(line);
+			break;
+		}
+	}
+	fclose(fp);
+}
+
+bool is_kernel_address(unsigned long addr)
+{
+	return addr >= kernel_base_addr;
+}
+
+unsigned long get_real_address(unsigned long addr)
+{
+	if (is_kernel_address(addr)) {
+		return (kernel_base_addr == 0x800000000000ULL) ?
+                    (addr | (-1UL << KADDR_SHIFT)) :
+                    (addr | kernel_base_addr);
+	}
+	return addr;
+}
```

Comment:
Oh, you're right.  But it reminds me of using 'unsigned long' for address is not safe for accessing 64-bit data on a 32-bit system. :-/

Anyway, I'd like to avoid the OR for 32-bit data.  What about this?

```
#include <limits.h>
...
unsigned long get_real_address(unsigned long addr)
{
        /* sign-extend to full 64-bit address */
        if (is_kernel_address(addr) && kernel_base_addr > UINT_MAX)
                    return addr | (-1ULL << KADDR_SHIFT);
        return addr;
}
```